File: problems/python/VendingMachineDemo/src/vending_machine_demo.py
"""Python demo equivalent of Java VendingMachineDemo."""
from enums.product_type import ProductType
from models.product import Product
from services.vending_machine import VendingMachine


def main() -> None:
    vending_machine = VendingMachine.get_instance()

    biscuit = Product(1, "Good-Day", 10.0, ProductType.BISCUITS)
    chocolate = Product(2, "Dairy Milk", 100.0, ProductType.CHOCOLATE)
    chips = Product(3, "Lays", 5.0, ProductType.CHIPS)

    vending_machine.load_product(1, biscuit, 3)
    vending_machine.load_product(2, chocolate, 15)
    vending_machine.load_product(3, chips, 5)

    vending_machine.display_products()

    print("****************************************")
    vending_machine.insert_money(20.0)
    vending_machine.select_product(3)
    print("****************************************")

    print("****************************************")
    vending_machine.insert_money(50.0)
    vending_machine.select_product(2)
    print("****************************************")

    print("****************************************")
    vending_machine.insert_money(5.0)
    vending_machine.select_product(3)
    print("****************************************")

    print("****************************************")
    vending_machine.insert_money(10.0)
    vending_machine.select_product(1)
    vending_machine.insert_money(20.0)
    vending_machine.select_product(1)
    vending_machine.insert_money(20.0)
    vending_machine.select_product(1)
    vending_machine.insert_money(10.0)
    vending_machine.select_product(1)
    print("****************************************")

    vending_machine.restock(1, 10)

    print("****************************************")
    vending_machine.insert_money(10.0)
    vending_machine.select_product(1)
    print("****************************************")


if __name__ == "__main__":
    main()

File: problems/java/VendingMachineDemo/src/VendingMachineDemo.java
import enums.ProductType;
import models.Product;
import services.VendingMachine;

public class VendingMachineDemo {
    public static void main(String[] args) {

        VendingMachine vendingMachine = VendingMachine.getInstance();

        // Machine has fixed racks capacity; operator only loads / restocks.
        // Operator: Creating products/items
        Product biscuit = new Product(1, "Good-Day", 10.0, ProductType.BISCUITS);
        Product chocolate = new Product(2, "Dairy Milk", 100.0, ProductType.CHOCOLATE);
        Product chips = new Product(3, "Lays", 5.0, ProductType.CHIPS);

        // Operator: loading products into machine
        vendingMachine.loadProduct(1, biscuit, 3);
        vendingMachine.loadProduct(2, chocolate, 15);
        vendingMachine.loadProduct(3, chips, 5);

        vendingMachine.displayProducts();

        System.out.println("****************************************");
        // Customer: Inserts Money (Extra money than product price)
        vendingMachine.insertMoney(20.0);
        vendingMachine.selectProduct(3);
        System.out.println("****************************************");

        System.out.println("****************************************");
        // Customer: Inserts Money (Less than product price)
        vendingMachine.insertMoney(50.0);
        vendingMachine.selectProduct(2);
        System.out.println("****************************************");

        System.out.println("****************************************");
        // Customer: Exact Payment (Equal to product price)
        vendingMachine.insertMoney(5.0);
        vendingMachine.selectProduct(3);
        System.out.println("****************************************");

        System.out.println("****************************************");
        // Customer: Depletes rack & retry (Until No quantity left)
        vendingMachine.insertMoney(10.0);
        vendingMachine.selectProduct(1);
        vendingMachine.insertMoney(20.0);
        vendingMachine.selectProduct(1);
        vendingMachine.insertMoney(20.0);
        vendingMachine.selectProduct(1);
        vendingMachine.insertMoney(10.0);
        vendingMachine.selectProduct(1);
        System.out.println("****************************************");

        // Operator: Restock (Increase the quantity of the product in the rack)
        vendingMachine.reStock(1, 10);

        System.out.println("****************************************");
        // Customer: Tries after restock (After restock, the quantity of the product in the rack is increased)
        vendingMachine.insertMoney(10.0);
        vendingMachine.selectProduct(1);
        System.out.println("****************************************");

    }
}
